### PR TITLE
7.1 Warshalls Algorithm

### LINKED ISSUE #21: US44: Implement Warshalls Algorithm (7.1)

### PR DESCRIPTION
**Pull request by usefuluselessness**
Takes an Input of either a adjacency matrix or adjacency relation

outputs of one the following:
adjacency matrix
adjacency relation
reachability matrix (through closure)
reachability matrix (through warshalls algorithm)

### CODE REVIEW

File: backend/solvers/warshall/Warshalls_Algorithm.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+def reachability_matrix_warshalls_algorithm(adjacency_matrix):
+    r = adjacency_matrix.copy()
+    matrix_size = len(adjacency_matrix)
+    for k in range(matrix_size):
+        #print_matrix(r, f"M_{k}:")
+        for i in range(matrix_size):
+            for j in range(matrix_size):
+
+                r[i][j] = r[i][j] or (r[i][k] and r[k][j])
+    #print_matrix(r, "R (Reachability Matrix):")
```
Comment (remirEstraven) [inline backend/solvers/warshall/Warshalls_Algorithm.py]: What are these commented out print statements for? And the other print statements? If they're just for personal testing, please remove them.

File: backend/solvers/tests/unit_tests/test_warshalls_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import pytest
+
+import sys
+from pathlib import Path
+sys.path.append(str(Path(__file__).resolve().parents[3]))
+
+from solvers.warshall import Warshalls_Algorithm
+
+# Textbook problem "Example 5"
+def test_reachability_matrix_through_closure():
```
Comment (remirEstraven) [inline backend/solvers/tests/unit_tests/test_warshalls_algorithm.py]: Good start on testing! I also really like comments explaining where the problems are from. Nice touch. Will need more tests for edge cases later, but I don't think that's in the scope of this PR. We can do that later.

File: backend/solvers/warshall/Warshalls_Algorithm.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
+﻿
+'''
+    Boolean matrix multiplication
+    Input: 2 (n x n matrix)
+    Output: n x n matrix
+
+    Boolean matrix multiplication:
+    matrix multiplication but uses logical AND for multiplication and logical OR for addition
+'''
+def boolean_multiply_matrices(A, B):
```
Comment (remirEstraven) [inline backend/solvers/warshall/Warshalls_Algorithm.py]: It'd be great if you could add type hints for these functions if possible like how Sam does [here](https://github.com/ssperry94/CS499-DiscreteMathSolver/blob/a93e067b1e5884436f6c7bd9490a1ee598d1d8e9/backend/solvers/logical_expression/wffparser.py#L14). I would harp on single letter variables, but for this math it makes sense.

File: backend/solvers/warshall/Warshalls_Algorithm.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    result = []
+    for t in data:
+        new_tuple = []
+        for x in t:
+            new_tuple.append(x - 1)
+        result.append(tuple(new_tuple))
+    return result
+
+# Example problem 5 and 6
+if __name__ == "__main__":
```
Comment (remirEstraven) [inline backend/solvers/warshall/Warshalls_Algorithm.py]: Same with this. If this is for local testing, please remove it. You will also need to add a call to `main.py` and `discrete_calculators.py` which will basically replace this.

**Review by remirEstraven — CHANGES_REQUESTED** [review]
Just a few small things that need some attention. Everything else looks great as far as I can tell. Good work!!! ^_^

File: backend/solvers/warshall/Warshalls_Algorithm.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    """
+    Prints A^1, A^2, ..., A^n (computed independently),
+    then prints R = A^1 OR A^2 OR ... OR A^n.
+    """
+    matrix_size = len(adjacency_matrix)
+
+    # A^1
+    Ak = [row[:] for row in adjacency_matrix]
+
+    #print_matrix(Ak, "A^1:")
```
Comment (ssperry94) [inline backend/solvers/warshall/Warshalls_Algorithm.py]: If we don't need this debugging statement you can go ahead and remove it

File: backend/solvers/warshall/Warshalls_Algorithm.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
+﻿
```
Comment (ssperry94) [inline backend/solvers/warshall/Warshalls_Algorithm.py]: I would add an empty __init__.py in the warshall directory as well so that anything we add under this directory is its own module

**Review by ssperry94 — CHANGES_REQUESTED** [review]

File: backend/solvers/tests/unit_tests/test_warshalls_algorithm.py (hunk 2 above)
Comment (usefuluselessness) [reply backend/solvers/tests/unit_tests/test_warshalls_algorithm.py]: I just added them to ensure my functions worked correctly.

File: backend/solvers/warshall/Warshalls_Algorithm.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+type number = int | float
+type matrix = List[List[int | float]]
+'''
+    Boolean matrix multiplication
+    Input: 2 (n x n matrix)
+    Output: n x n matrix
+
+    Boolean matrix multiplication:
+    matrix multiplication but uses logical AND for multiplication and logical OR for addition
+'''
```
Comment (ssperry94) [inline backend/solvers/warshall/Warshalls_Algorithm.py]: @baileylitton3 are we enforcing docstrings, or is this ok?

File: backend/solvers/section_7_1_warshalls_algorithm/warshalls_algorithm.py
Comment (ssperry94) [inline backend/solvers/section_7_1_warshalls_algorithm/warshalls_algorithm.py:1]: do you think you'll add a function for warshall's in discrete_calculators, or do that in a future issue?

**Review by ssperry94 — APPROVED** [review]
Looks good overall, just a few questions is all.

**Review by ssperry94 — APPROVED** [review]

**Review by remirEstraven — APPROVED** [review]
